Add the demo run to the website pages (#428)
ImperialCollegeLondon/virtual_ecosystem: docs/source/virtual_ecosystem/virtual_ecosystem_in_use.md

---
jupytext:
  formats: md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.13.8
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
settings:
  output_matplotlib_strings: remove
---

# The Virtual Ecosystem model

The code below is a brief demonstration of the Virtual Ecosystem model in operation.
The workflow of the model is:

## Create the model configuration and initial data

> Here we are using the example data files supplied with the `virtual_ecosystem`

Reviewer: ``
```suggestion
Here we are using the example data supplied with the `virtual_ecosystem`
```

package, which supplies a set of example data files and a simple model configuration
to run a simulation. The following command line arguments set up the example data
directory in Linux, Mac or Windows Subsystem for Linux (WSL).

```{code-cell}
import pathlib

import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import numpy as np
import xarray
```

```{code-cell}
:tags: [remove-cell]

%%bash
# Remove any existing VE data directory in the /tmp/ directory
if [ -d /tmp/ve_example ]; then
  rm -r /tmp/ve_example
fi
```

```{code-cell}
%%bash
# Install the example data directory from the Virtual Ecosystem package
ve_run --install-example /tmp/
```

> The `ve_example` directory contains the following files. These include:

Reviewer: ```suggestion
The `ve_example` directory contains the following files:
```

    --progress \
```

The log file is very long and shows the process of running the model. The code below
shows the start and end lines from the log to give and idea of what it contains.

```{code-cell}
# Open and read the log
with open("/tmp/ve_example/out/logfile.log") as log:
    log_entries = log.readlines()

# Print the first lines
for entry in log_entries[:6]:
    print(entry.strip())

print("...")

# Print the last lines
for entry in log_entries[-5:]:
    print(entry.strip())
```

## Looking at the results

The Virtual Ecosystem writes out a number of data files:

* `initial_state.nc`: A single compiled file of the initial input data.
* `all_continuous_data.nc`: An optional record of time series data of the variables
  updated at each time step.
* `final_state.nc`: The model data state at the end of the final step.

These files are written to the standard NetCDF data file format.

```{code-cell}
# Load the generated data files
initial_state = xarray.load_dataset("/tmp/ve_example/out/initial_state.nc")
continuous_data = xarray.load_dataset("/tmp/ve_example/out/all_continuous_data.nc")
final_state = xarray.load_dataset("/tmp/ve_example/out/final_state.nc")
```

> ### Initial state

Reviewer: ```suggestion
### Initial state and input data
```
You show the climate input time series here so I think we should include this in the title

> The `initial_state.nc` file contains all of the data required to start up the model.

Reviewer: Add something like: It also contains time series of climate variables at reference height that drive the model.

# Elevation
im1 = ax1.imshow(initial_state["elevation"].to_numpy().reshape((9, 9)), extent=extent)
ax1.set_title("Elevation (m)")
fig.colorbar(im1, ax=ax1, shrink=0.7)

# Initial soil carbon
im2 = ax2.imshow(initial_state["pH"].to_numpy().reshape((9, 9)), extent=extent)
ax2.set_title("Soil pH (-)")
fig.colorbar(im2, ax=ax2, shrink=0.7)

plt.tight_layout();
```

For some variables, it may be useful to visualise spatial structure in 3 dimensions.
The obvious candidate is elevation.

```{code-cell}
# Extract the elevation data for a 3D plot
top = initial_state["elevation"].to_numpy()
x = continuous_data["x"].to_numpy()
y = continuous_data["y"].to_numpy()
bottom = np.zeros_like(top)
width = depth = 90
```

```{code-cell}
# Make a 3D barplot of the elevation
fig = plt.figure(figsize=(10, 8))
ax = fig.add_subplot(projection="3d")
colors = plt.cm.turbo(top.flatten() / float(top.max()))

poly = ax.bar3d(x, y, bottom, width, depth, top, shade=True, color=colors)
ax.set_title("Elevation (m)")

cell_bounds = range(0, 811, 90)
ax.set_xticks(cell_bounds)
ax.set_yticks(cell_bounds);
```

For other variables, such as air temperature and precipitation, the initial data

> also provides time series data across the grid cells that are used to force the

Reviewer: ```suggestion
also provides time series data at reference height that are used to force the
```

simulation across the configured time period.

```{code-cell}
initial_state
```

```{code-cell}
# Make two side by side plots
fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(12, 5))

# Air temperature
ax1.plot(initial_state["time_index"], initial_state["air_temperature_ref"])

> ax1.set_title("Air temperature across grid cells")

Reviewer: ```suggestion
ax1.set_title("Air temperature at reference height")
```

ax1.set_ylabel("Air temperature (°C)")
ax1.set_xlabel("Time step (months)")

# Precipitation
ax2.plot(initial_state["time_index"], initial_state["precipitation"])

> ax2.set_title("Precipitation across grid cells")

Reviewer: ```suggestion
ax2.set_title("Precipitation forcing")
```
